### PR TITLE
Fix Trade Republic PDF-Importer

### PR DESCRIPTION
**Pull request by Nirus2000**
Fix Trade Republic PDF importer after comments in #3981

Hello @christen90 
[> I can't provide a debug of something which is not contained. You can see in both exports an interest booking. Both do not contain taxes.](https://github.com/portfolio-performance/portfolio/pull/3981#discussion_r1589190432)

I would leave the interest postings as they are, as they are already successfully implemented in addAccountStatementTransaction_Format01() (older transactions).

---

[> missing 16.4.2024 18,34 ALIEXPRESS.COM 16.4.2024 17,53 ALIEXPRESS.COM 16.4.2024 19,40 Lidl sagt Danke 16.4.2024 0,08 Visa Geld zurueck Aktion](https://github.com/portfolio-performance/portfolio/pull/3981#discussion_r1589052088)

Done.... thx 👍🏻 

---

[> Saldo gets increased --> deposit](https://github.com/portfolio-performance/portfolio/pull/3981#discussion_r1589052976)

Done.... thx 👍🏻 
I have changed the behavior for the transaction type "card transaction" and added an additional function that checks whether the amount in the previous transaction is larger or smaller.

---

[> what if user generates statement from Dec 2023 to Jan 2024?](https://github.com/portfolio-performance/portfolio/pull/3981#discussion_r1589059421)

As soon as we have an example here, we can change this. I have already prepared the extractor for this and adapted the regEx for this. However, I believe that this will not happen because the document always has a fixed period. Let's hope for the best... 👍🏻 

---

[ how to catch other paybacks, refunds?
 See aliexpress booking on 17th Apr. Saldo gets increased compared to previous row.](https://github.com/portfolio-performance/portfolio/pull/3981#discussion_r1589184902)

I think now it works 👍🏻 

---

[other format:
 30 Apr.
 2024 Überweisung
 Einzahlung akzeptiert: DE00000000000000000000 auf
 DE00000000000000000000 1.200,00 € 18.514,32 €](https://github.com/portfolio-performance/portfolio/pull/3981#discussion_r1589064416)

Done... 👍🏻 

You can check the whole thing and if something is still wrong, please let me know... Thank you again.... !!!

### CODE REVIEW

**Comment by christen90** [comment]
Example covering two years:
[Kontoauszug08.txt](https://github.com/portfolio-performance/portfolio/files/15201938/Kontoauszug08.txt)

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java [hunk 1]
```diff
@@ -1402,19 +1435,13 @@ private void addAccountStatementTransaction_Format02()
                                         // DE00000000000000000000 1.200,00 € 18.514,32 €
```
Comment (christen90) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java:1435]: between Überweisung and Einzahlung no line break (see Kontoauszug08.txt:89)

**Review by christen90 — CHANGES_REQUESTED** [review]
Hi Alex,
looks great. 🥳 I like the helper function to obtail row above.

Regarding taxes of interests I would recomment to mark also in "old" format as nonImportable. They also lack of taxual details. The Abrechnung of interest at beginning of month already have the info available.

Also a remark... It's not old and new... The old one is automatically created each quarter. The new one can be manually triggered (and range can be specified on monthly basis):
image

BR Christian

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java [hunk 2]
```diff
@@ -2352,6 +2379,48 @@ private <T extends Transaction<?>> void addFeesSectionsTransaction(T transaction
...
+
+    private static class AccountAmountTransactionHelper
+    {
+        private List<AccountAmountTransactionItem> items = new ArrayList<>();
+
+        // Finds an AccountAmountTransactionItem in the list that has a line
+        // number less than or equal to the specified line
+        public Optional<AccountAmountTransactionItem> findItem(Integer line, Money money)
+        {
+            if (items.isEmpty())
```
[the remark this answers is not in the document]
Comment (Nirus2000) [reply name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java:2404]: With this patter
`Pattern pAccountAmountTransaction = Pattern.compile("^.* [\\.,\\d]+ \\p{Sc} (?<amount>[\\.,\\d]+) (?<currency>\\p{Sc})$");`
we get all amounts in the document that correspond to the syntax. This is therefore the last one to be used. :stuck_out_tongue_winking_eye:
Theoretically, we can also leave it out, but for completeness and security, we use it.

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java [hunk 3]
```diff
@@ -1327,12 +1333,31 @@ private void addAccountStatementTransaction_Format02()
         final DocumentType type = new DocumentType("KONTO.BERSICHT", //
                         documentContext -> documentContext //
                                         // @formatter:off
-                                        // Erstellt am 29 Apr. 2024
+                                        // aeJkGYj LzyOwoZH DATUM 01 Apr. 2024 - 28 Apr. 2024
                                         // @formatter:on
                                         .section("year") //
-                                        .match("^Erstellt am [\\d]{2} .* (?<year>[\\d]{4})$") //
```
[the remark this answers is not in the document]
Comment (Nirus2000) [reply name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java:1333]: with:
```
documentContext -> documentContext //
                                        // @formatter:off
                                        // aeJkGYj LzyOwoZH DATUM 01 Apr. 2024 - 28 Apr. 2024
                                        // @formatter:on
                                        .section("year") //
                                        .match("^.* DATUM [\\d]{2} .* [\\d]{4} \\- [\\d]{2} .* (?<year>[\\d]{4})$") //
                                        .assign((ctx, v) -> ctx.put("year", v.get("year"))), //
```
We intercept the year string and check later in the "sections"
```
                        .section("date", "amount", "currency") //
                        .documentContext("year") // <---------- HERE HERE HERE
                        .match("^(?<date>[\\d]{2} [\\w]{3,4}([\\.]{1})?) .berweisung Einzahlung akzeptiert:.* (?<amount>[\\.,\\d]+) (?<currency>\\p{Sc}) [\\.,\\d]+ \\p{Sc}$") //
                        .assign((t, v) -> {
                            t.setDateTime(asDate(v.get("date") + " " + v.get("year")));
                            t.setAmount(asAmount(v.get("amount")));
                            t.setCurrencyCode(asCurrencyCode(v.get("currency")));
                        })
```
 whether it is valid... if not, we have an error

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (christen90) [reply name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java:2404]: ANFANGSSALDO isn't on same position as on the other transactions.
^Depotkonto (?<anfangssaldo>[\\.,\\d]+) \\p{Sc} [\\.,\\d]+ \\p{Sc} [\\.,\\d]+ \\p{Sc} [\\.,\\d]+ \\p{Sc}$

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java [hunk 4]
```diff
@@ -1379,16 +1404,24 @@ private void addAccountStatementTransaction_Format02()
...
+                                                            Money amountAfter = Money.of(asCurrencyCode(v.get("currencyAfter")), asAmount(v.get("amountAfter")));
+
+                                                            AccountAmountTransactionHelper accountAmountTransactionHelper = context.getType(AccountAmountTransactionHelper.class).orElseGet(AccountAmountTransactionHelper::new);
+                                                            Optional<AccountAmountTransactionItem> item = accountAmountTransactionHelper.findItem(v.getStartLineNumber(), amountAfter);
+
+                                                            if (item.isPresent())
+                                                            {
+                                                                Money amountBefore = Money.of(item.get().currency, item.get().amount);
+
+                                                                if (amountBefore.isGreaterThan(amountAfter))
```
[the remark this answers is not in the document]
Comment (christen90) [reply name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java:1422]: Thanks for the hint, what I was trying to say is, that an equation is more exact that a greater than.

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Nirus2000) [reply name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java:2404]: Ah... Okay. I'll see it. 👍🏻

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Nirus2000) [reply name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java:1422]: We do not want to check for equality, but we want to check whether the balance of the previous transaction is smaller than the balance afterwards. 😜

**Comment by Nirus2000** [comment]
Recheck again @christen90

**Comment by christen90** [comment]
looks even better

only finding   name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TradeRepublicPDFExtractor.java:1481
      // 024 Gebühren Trade Republic Card 5,00 € 49.997,41 €
should be 2024...

I still would skip the interest bookings.

**Pull request merged**